src/main/java/com/example/demo/repositories/UserRepository.java

```diff
@@ -0,0 +1,154 @@
+package com.example.demo.repositories;
+
+import com.example.demo.models.User;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import javax.sql.DataSource;
+import java.util.List;
+import java.util.Optional;
+
+
+@Repository
+public class UserRepository {
+
+    private final Logger logger = LoggerFactory.getLogger(UserRepository.class);
+
+    private final JdbcTemplate template;
+    private final NamedParameterJdbcTemplate namedTemplate;
+
+    @Autowired
+    public UserRepository(DataSource dataSource) {
+        this.template = new JdbcTemplate(dataSource);
+        this.namedTemplate = new NamedParameterJdbcTemplate(dataSource);
+    }
+
+    public List<User> getAllUsers() {
+
+        String query = """
+                SELECT id, first_name, last_name, email FROM users
+                                            WHERE email LIKE :emailFilter
+                """;
+
+        MapSqlParameterSource paramMap = new MapSqlParameterSource() {{
+            addValue("emailFilter", "%.com");
+        }};
+
+        try {
+            return namedTemplate.query(query, paramMap, new BeanPropertyRowMapper<>(User.class));
+        } catch (Exception e) {
+            logger.warn("Error getting all users - e={}", e.toString());
+            throw e;
+        }
+    }
+
+    public Optional<User> findById(String id) {
+        String query = """
+                SELECT * FROM users WHERE id = :idFilter
```

Comment:
Let's change this to be a list of the columns we want. This prevents breaking this endpoint if we update the db.

Reply:
Do you mean to use a list instead of "*" in the select statement? 
What would that look like and how would that be resistant to changes in the DB? Wouldn't you need to go in and manually change the columns in the list each time you changed the DB? I'd just like to understand what problems might arise if we were to keep it like this. 
I'm guessing this would cause issues if we ever had data the user shouldn't see and maybe this is the sort of change that you're referring to.

Reply:
Yes, I mean a list of columns. 

It would break the API entirely as soon as the DB was updated with a new column. This is because the beanproperty row mapper would attempt to deserialize the new columns, and throw an exception because it would not be able to find the correlating properties in the object. 

A very important principle in all coding is to maintain backwards compatibility to prevent breaking features customers may rely on.

Reply:
In the case of adding a new column to a table that already has rows, is there a way to provide a default value for the rows that weren't initialized with a value?
Also, if you are updating the table, then you'd also be updating the class. Isn't there something that could be done to account for 'no value' when the row mapper is doing it's thing? 
Is it an issue because when you add a column to a table, the rows in the table don't get any sort of update at all? Not even a null value appears, that cell is simply empty, which leads to an error that can't be handled?

Reply:
Yes.
Yes, typically so. However, there will be other devs working in the same repo, and they don't need your permission to make DB changes. Not are they required to make you aware of them. It is better to air on the side of caution and make your code as robust as possible.

Yes, you can also provide default values on the class.

No. The error is caused because the mapper presumes all data being returned from the query is directly mappable to properties in the class. When it can't find the property for a column, it will break.

Reply:
Alright, so the moral of the story: 

When writing out SQL statements, do it in such a way that changes in the DB aren't going to cause my code to break. These DB changes could arise from a natural progression of the system and keeping future changes in mind will help to keep your code backward compatible. Also, since multiple collaborators could be working on this system at the same time, you want to write code that is resistant to the potential changes of others.

Is that about right? 

So far, this discussion seems to rotate around the idea of new columns being added. 

What considerations should be made when removing a column or changing a column? These sorts of changes seem more likely to cause issues with code and seem to be harder to handle if your goal is future proofing or robustness.

Reply:
Yes, that's right. 

For your examples, there isn't much you can do to remediate those. :/ If someone else removes a column or changes its type, your toast. 😅